presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java

```diff
@@ -143,16 +162,41 @@ public Iterable<HivePartition> getPartitionsIterator(
         else {
             return () -> {
                 List<String> partitionNames = partitionFilteringFromMetastoreEnabled ? getFilteredPartitionNames(session, metastore, hiveTableHandle, effectivePredicate) : getAllPartitionNames(session, metastore, hiveTableHandle, constraint);
-                return partitionNames.stream()
-                        // Apply extra filters which could not be done by getFilteredPartitionNames
-                        .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, constraint))
-                        .filter(Optional::isPresent)
-                        .map(Optional::get)
-                        .iterator();
+
+                if (isParallelPartitionValueParsingEnabled(session) && partitionNames.size() > PARTITION_NAMES_BATCH_SIZE) {
+                    List<List<String>> partitionBatches = Lists.partition(partitionNames, PARTITION_NAMES_BATCH_SIZE);
+                    // Use CopyOnWriteArrayList to prevent race condition when multiple threads try to add partitions to this list
+                    List<HivePartition> result = new CopyOnWriteArrayList<>();
```

Comment:
Hmm. Did concurrent linked list or such perform worse ? Is it possible to make this be a list of batches so that the add is trivial (instead of addAll)

Reply:
+1 seems like concurrent linked list should be good enough. CopyOnWriteArrayList should be required for read heavy use cases.